### PR TITLE
Volume encryption support

### PR DESCRIPTION
**Pull request by imran-ansari**
Volume encryption support is a host-based encryption using LUKS tool called cryptsetup. During node-stage flow, cryptsetup tool is used on the discovered device to enable encryption. For details about the approach please refer to https://rndwiki.corp.hpecorp.net/confluence/display/storageecosystem/Volume+encryption+support+in+CSI+Driver

@rgcostea , @raunakkumar , @wdurairaj , @sneharai4 - please review the changes

### CODE REVIEW

File: linux/device.go [hunk 1]
```diff
@@ -495,6 +515,54 @@ func createLinuxDevice(volume *model.Volume) (dev *model.Device, err error) {
...
+						return nil, err
+					}
+
+					// LUKS format device if this is the first time it is being used
+					if !isLuksDev {
+						log.Infof("Device %s is a new device. LUKS formatting it...", originalDevPath)
+						formatCmd := exec.Command("cryptsetup", "luksFormat", "--batch-mode", originalDevPath)
+						log.Debugf("luksFormat command:%s", formatCmd)
+						formatCmd.Stdin = strings.NewReader(volume.EncryptionKey)
+						if err := formatCmd.Run(); err != nil {
```
Comment (imran-ansari) [inline linux/device.go:533]: util.ExecCommandOutput() should have been used here but cryptsetup requires reading passphrase from stdin which at the moment is not supported by util.ExecCommandOutput(). This is a work in progress.

Comment (raunakkumar) [reply linux/device.go:533]: is there a way to pass it as an argument ?

File: linux/device.go [hunk 2]
```diff
@@ -1234,3 +1315,37 @@ func ExpandDevice(targetPath string, volAccessType model.VolumeAccessType) error
...
+	log.Debugf(">>>> resizeMappedLuksDevice - %s", devPath)
+	defer log.Debugf("<<<< resizeMappedLuksDevice - %s", devPath)
+
+	_, exitStatus, _ := util.ExecCommandOutput("cryptsetup", []string{"resize", devPath})
+	log.Debugf("exit status of resize LUKS device: %d", exitStatus)
+	if exitStatus == 0 {
+		log.Infof("mapped LUKS device %s resized successfully", devPath)
+		return true, nil
+	}
+	msg := fmt.Errorf("failed to resize mapped LUKS device %s", devPath)
```
Comment (raunakkumar) [inline linux/device.go:1329]: nit: add reason for failed resize

File: linux/device.go [hunk 3]
```diff
@@ -1234,3 +1315,37 @@ func ExpandDevice(targetPath string, volAccessType model.VolumeAccessType) error
...
+	log.Debugln(msg.Error())
+		return false, msg
+}
+
+func isMappedLuksDevice(devPath string) (bool, error) {
+	log.Debugf(">>>> isMappedLuksDevice - %s", devPath)
+	defer log.Debugf("<<<< isMappedLuksDevice - %s", devPath)
+
+	_, exitStatus, _ := util.ExecCommandOutput("cryptsetup", []string{"status", "-v", devPath})
```
Comment (raunakkumar) [inline linux/device.go:1338]: ExecCommandOutput returns stdout and stderr in a single string, the return code, and error.
Can we log some more information from the command output ?

File: linux/mount.go [hunk 4]
```diff
@@ -914,10 +928,13 @@ func SetupFilesystem(device *model.Device, filesystemType string) error {
 		return nil
 	}
 
-	log.Tracef("Creating filesystem %s on device path %s", filesystemType, device.AltFullPathName)
-	if err := RetryCreateFileSystem(device.AltFullPathName, filesystemType); err != nil {
-		log.Errorf("Failed to create filesystem %s on device with path %s", filesystemType, device.AltFullPathName)
-		return err
+	// For clone case, filesystem will be present
+	if !device.FilesystemPresent {
```
Comment (raunakkumar) [inline linux/mount.go:932]: This flag is set only when EncryptionKey is set? What about the existing scenario where `volume.EncryptionKey == ""` This needs to handle both the cases.

**Review by raunakkumar — CHANGES_REQUESTED** [review]
Please address the review comments and update the log levels to trace.

File: linux/device.go (hunk 1 above)
Comment (imran-ansari) [reply linux/device.go:533]: The only way to pass it as an argument seems to be via a file containing the passphrase. However, it seems to be an overkill to create an temporary file on the fly and then to delete it once the job is done.

File: linux/device.go [hunk 5]
```diff
@@ -467,6 +468,25 @@ func rescanLoginVolumeForBackend(volObj *model.Volume) error {
 	return nil
 }
 
+func isLuksDevice(devPath string) (bool, error) {
+	log.Debugf(">>>> isLuksDevice - %s", devPath)
```
[the remark this answers is not in the document]
Comment (imran-ansari) [reply linux/device.go:472]: okay

File: linux/device.go [hunk 6]
```diff
@@ -467,6 +468,25 @@ func rescanLoginVolumeForBackend(volObj *model.Volume) error {
...
+		return true, nil
+	}
+	if exitStatus == 1 {
+		log.Debugf("Not a LUKS device - %s", devPath)
+		return false, nil
+	} else {
+		log.Debugf("Unknown device - %s", devPath)
+		return false, fmt.Errorf("isLuks command failed to find out if the device %s is encrypted", devPath)
+	}
+	//return strings.EqualFold(out, "Command successful."), nil
```
[the remark this answers is not in the document]
Comment (imran-ansari) [reply linux/device.go:487]: okay

File: linux/device.go (hunk 3 above)
Comment (imran-ansari) [reply linux/device.go:1338]: sure

File: linux/device.go (hunk 2 above)
Comment (imran-ansari) [reply linux/device.go:1329]: sure

File: linux/device.go [hunk 7]
```diff
@@ -495,6 +515,54 @@ func createLinuxDevice(volume *model.Volume) (dev *model.Device, err error) {
...
+
+					isLuksDev, err := isLuksDevice(originalDevPath)
+					if err != nil {
+						return nil, err
+					}
+
+					// LUKS format device if this is the first time it is being used
+					if !isLuksDev {
+						log.Infof("Device %s is a new device. LUKS formatting it...", originalDevPath)
+						formatCmd := exec.Command("cryptsetup", "luksFormat", "--batch-mode", originalDevPath)
```
[the remark this answers is not in the document]
Comment (imran-ansari) [reply linux/device.go:530]: cyrptsetup is being bundled up with the CSI-Driver image - so I believe we should be good here.
Please check out the following Dockerfile diff:
https://github.com/hpe-storage/csi-driver/pull/246/files#diff-3cc87feca0c7dd59008a384ebb2d5945411952eb21e58166f8b970a321661a48

File: linux/mount.go (hunk 4 above)
Comment (imran-ansari) [reply linux/mount.go:932]: This flag is not required anymore. It was added to disallow mkfs on the cloned-volume. However, after the lines 916-919 were added, this flag has become redundant:

https://github.com/hpe-storage/common-host-libs/blob/6ab488305b0c4725ea9a873a4ceaf2066ee7cfab/linux/mount.go#L916

**Comment by imran-ansari** [comment]
> 
> 
> Please address the review comments and update the log levels to trace.

Requested changes have been addressed. Please have a look.

File: linux/device.go [hunk 8]
```diff
@@ -495,6 +508,49 @@ func createLinuxDevice(volume *model.Volume) (dev *model.Device, err error) {
...
+
+					//finding major minor for the encrypted path
+					//args := []string{"ls", "--target", "crypt" "|",encryptedPath}
+					//out, _, err := util.ExecCommandOutput(dmsetupcommand, args)
+					//if err != nil {
+					//      return nil, fmt.Errorf("failed to retrieve crypt device with encryptedPath %s : %s", //encryptedPath, err.Error())
+					//}
+					//return d, nil
+				}
+				log.Debugf("after encryption device with matching SerialNumber:%s map %s and slaves %+v", d.SerialNumber, d.AltFullPathName, d.Slaves)
```
Comment (raunakkumar) [inline linux/device.go]: nit: after encrypting device

File: linux/device.go [hunk 9]
```diff
@@ -495,6 +508,49 @@ func createLinuxDevice(volume *model.Volume) (dev *model.Device, err error) {
...
+					log.Infof("luksOpen command:%s",openCmd)
+					openCmd.Stdin = strings.NewReader(volume.EncryptionKey)
+					if err := openCmd.Run(); err != nil {
+						log.Errorf("Open command failed with error %v", err)
+					}
+					// Replacing the device path,AltFullPathName
+					d.LuksPathname = encryptedPath
+					d.AltFullLuksPathName = "/dev/mapper/" + encryptedPath
+
+					//finding major minor for the encrypted path
```
Comment (raunakkumar) [inline linux/device.go]: remove the commented code

**Review by raunakkumar — CHANGES_REQUESTED** [review]

File: linux/device.go (hunk 9 above)
Comment (imran-ansari) [reply linux/device.go]: @raunakkumar - this seems to be an old code which is not there in the PR anymore.
 Please check the commit:  https://github.com/hpe-storage/common-host-libs/pull/143/commits/9a214117f265b436fea563170605268fb2b5d549

and the modified file in my repo:
https://github.com/imran-ansari/common-host-libs/blob/vol_encryption/linux/device.go

File: linux/device.go (hunk 8 above)
Comment (imran-ansari) [reply linux/device.go]: @raunakkumar - this seems to be an old code which is not there in the PR anymore.
 Please check the commit:  https://github.com/hpe-storage/common-host-libs/pull/143/commits/9a214117f265b436fea563170605268fb2b5d549

and the modified file in my repo:
https://github.com/imran-ansari/common-host-libs/blob/vol_encryption/linux/device.go

**Review by raunakkumar — APPROVED** [review]
lgtm

**Pull request merged**